### PR TITLE
[SPARK-47413][SQL] - add support to substr/left/right for collations  

### PR DESCRIPTION
**Pull request by GideonPotok**
https://issues.apache.org/jira/browse/SPARK-46830

### What changes were proposed in this pull request?

Add collation support to types of return values for calls to substr, left, right, when passed in arguments of an explicit, implicit, or session-specified collations. Add tests to validate behavior. 

### Why are the changes needed?

We are incrementally adding collation support to built-in string functions in Spark. These functions are intended to be supported for collated types. 

### Does this PR introduce _any_ user-facing change?

these sql functions will now not throw errors when passed in collated types. Instead, they will return the right value, of the passed in type. Or of the default collation.

### How was this patch tested?

Unit testing + ad-hoc spark shell and pyspark shell interactions. 

### Was this patch authored or co-authored using generative AI tooling? 

No.

### CODE REVIEW

**Comment by GideonPotok** [comment]
@uros-db  

I have some tests, and have substring implementation basically passing tests.
However, my redefined implementations of Left and Right are failing my tests. They are currently throwing the following exceptions in my new test cases in CollationSuite.scala:

```
[COMPLEX_EXPRESSION_UNSUPPORTED_INPUT.MISMATCHED_TYPES] Cannot process input data types for the expression: "(IF((1 <= 0), , substring(collate(klm), (- 1), 2147483647)))". All input types must be the same except nullable, containsNull, valueContainsNull flags, but found the input types ["STRING", "STRING COLLATE UTF8_BINARY_LCASE"]. SQLSTATE: 42K09
```

I think it is caused by additional work needed on the overridden implementation of `replacement` within `Left` and `Right` in `stringExpressions.scala`. I think that because the second parameter to Literal is DataType in the replacement, rather than AbstractDataType, there is this little issue. I think that is what is causing these tests to fail. What do you think? Please advise on how to debug this issue.

**Comment by uros-db** [comment]
since the "COMPLEX_EXPRESSION_UNSUPPORTED_INPUT.MISMATCHED_TYPES" problem seems to be in `Right.replacement` expression when doing `Literal(value, dataType)`, did you try something like `Literal(UTF8String.EMPTY_UTF8, str.dataType)` to match these datatypes?

**Comment by GideonPotok** [comment]
@uros-db I think that did the trick! I will let you know when to review again. Thanks!

File: sql/core/src/test/scala/org/apache/spark/sql/CollationStringExpressionsSuite.scala
Comment (uros-db) [inline sql/core/src/test/scala/org/apache/spark/sql/CollationStringExpressionsSuite.scala:92]: try covering some more edge cases in these tests, such as: empty strings, uppercase and lowercase mix, different byte-length chars, etc.

good example would be this PR: https://github.com/apache/spark/pull/45749

**Comment by uros-db** [comment]
also, use `CollationStringExpressionsSuite` instead of `CollationSuite` for these tests

File: sql/core/src/test/scala/org/apache/spark/sql/CollationStringExpressionsSuite.scala
Comment (GideonPotok) [reply sql/core/src/test/scala/org/apache/spark/sql/CollationStringExpressionsSuite.scala:92]: @uros-db Please see what I have since added and let me know whether what to keep going, or whether we have enough!

Comment (uros-db) [reply sql/core/src/test/scala/org/apache/spark/sql/CollationStringExpressionsSuite.scala:95]: I think this will do it, thanks! Now just move over all tests to `CollationStringExpressionsSuite`, re-run any failing CI checks, and modify the PR title so we can open this up for review to Spark committers

Comment (GideonPotok) [reply sql/core/src/test/scala/org/apache/spark/sql/CollationStringExpressionsSuite.scala:95]: Sorry @uros-db, I didn't realize you wanted me to add the additional edge cases to this suite specifically. Are you sure you want me to move _all_ the new tests to `CollationStringExpressionsSuite`?

Comment (uros-db) [reply sql/core/src/test/scala/org/apache/spark/sql/CollationStringExpressionsSuite.scala:95]: yeah, let's put all these tests that are specific to string functions into the separate suite `CollationStringExpressionsSuite`

this way, `CollationSuite` won't get cluttered as we continue adding more and more support for various string expressions

**Comment by GideonPotok** [comment]
@uros-db this would be ready for review but spark sql unit tests repeatedly fail for CSVLegacyTimeParserSuite, which I did not modify. I have reran it a total of four times. Any idea how to get this check to pass successfully?

**Comment by uros-db** [comment]
did you try to run this suite locally and investigate any potential issues?

**Comment by uros-db** [comment]
I don't think `CSVLegacyTimeParserSuite` is related to you, but it would probably be a very good idea to setup Maven so that you can run/debug all tests locally in general

**Comment by GideonPotok** [comment]
> I don't think `CSVLegacyTimeParserSuite` is related to you, but it would probably be a very good idea to setup Maven so that you can run/debug all tests locally in general

Same issue running tests locally with maven. I always build with maven and when running tests with it (Eg `build/mvn -Dtest=none -DwildcardSuites=org.apache.spark.sql.execution.datasources.csv.CSVLegacyTimeParserSuite test`), i encounter the same hanging on anything that is writing to file system. I can keep looking into it.

I agree that my changes are probably not what is causing `org.apache.spark.sql.execution.datasources.csv.CSVLegacyTimeParserSuite` to fail in GHA.

**Comment by GideonPotok** [comment]
@uros-db I got the file-writing tests to work locally when I simply `export SPARK_HOME=/Users/gideon/repos/spark` prior to running my maven tests. 

More importantly, All GHA are passing. So this is ready for reviewers to begin to review.

**Comment by GideonPotok** [comment]
@uros-db this is ready for review.

**Comment by uros-db** [comment]
@GideonPotok nice work, thanks!

Heads up though: we will soon be finishing some code refactoring related to collation-aware string expression support ([SPARK-47410](https://issues.apache.org/jira/browse/SPARK-47410)), and will likely need to rewrite this PRs a bit in order to comply with some new design before proceeding to final reivew

let's put this PR on hold for now, and I'll ping you when we're ready to move on

**Comment by GideonPotok** [comment]
@uros-db No problem at all.

 if I understand your refactor correctly, my changes will basically either stay in the same place or move to the new `common/unsafe/src/main/java/org/apache/spark/sql/catalyst/util/CollationSupport.java`, right? 

I will in the meantime take care of implementing https://issues.apache.org/jira/browse/SPARK-47412 and will also then put that on hold until after the refactor is merged. If that sounds good?

**Comment by GideonPotok** [comment]
PS: Do you think changes, such as these, which are only to implementations of `inputTypes` and `replacement`, which do not rely on calling UTFString or CollationFactory, will need to be modified? Even if they won't, it makes sense for me to hold off.  But I would still like to know how you see this changing.

Thank you, kind sir :)

**Comment by uros-db** [comment]
@GideonPotok You are correct, this refactor should not greatly affect your current PR in particular - I expect you'll only need to refactor testing a bit (shouldn't be too much work)

Feel free to move over to the next ticket, and I'll let you know when and how to consolidate your code/design so we can move to final review and sign off on this PR

**Review by uros-db — CHANGES_REQUESTED** [review]
we’ve done some major code restructuring in https://github.com/apache/spark/pull/45978, so please sync these changes before moving on

@GideonPotok you’ll likely need to rewrite the code in this PR a bit, so please follow the guidelines outlined in https://issues.apache.org/jira/browse/SPARK-47410

**Comment by GideonPotok** [comment]
@ur.

**Pull request closed without merging**
